File: java/semaphori/src/main/java/org/example/Producer.java
package org.example;

public class Producer implements Runnable {
    private int numItemsToProduce;
    private Storage storage;
    private int producerId;
    private Thread thread;

    public Producer(int producerId, Storage storage, int numItemsToProduce) {
        this.numItemsToProduce = numItemsToProduce;
        this.storage = storage;
        this.producerId = producerId;

        this.thread = new Thread(this);
        this.thread.start();
    }

    @Override
    public void run() {
        for (int i = 0; i < numItemsToProduce; i++) {
            storage.produceItem(producerId, i);
        }
        System.out.println("Producer " + producerId + " finished production.");
    }
}

File: c#/semaphori/semaphori/Producer.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading;
using System.Threading.Tasks;

namespace semaphori
{
    public class Producer
    {
        private int numItemsToProduce;
        private Storage storage;
        private int producerId;

        public Producer(int producerId, Storage storage, int numItemsToProduce)
        {
            this.numItemsToProduce = numItemsToProduce;
            this.storage = storage;
            this.producerId = producerId;

            Thread workerThread = new Thread(RunProduction);
            workerThread.Start();
        }

        private void RunProduction()
        {
            for (int i = 0; i < numItemsToProduce; i++)
            {
                storage.ProduceItem(producerId, i);
            }
            Console.WriteLine($"Producer {producerId} finished production.");
        }
    }
}